src/view-backend-exportable-private.h

```diff
@@ -102,6 +102,11 @@ class ViewBackend final : public WS::ExportableClient, public FdoIPC::MessageRec
     int m_clientFd { -1 };
 };
 
+struct wl_client_destroy_listener {
+    ViewBackend* backend;
+    struct wl_listener destroyClientListener;
```

Comment:
Is there any reason why the `wl_listener` is not a member of the `ViewBackend` type?

Reply:
This was my first try but I got in crashes in situations where the ViewBackend is destroyed before the wl_client client is deleted. That is the normal case where the ViewBackend is destroyed because the WebProcess is not longer needed (elimination of all the views related to one WebContext). In that scenario when the wl_callback for the listener is called the memory where the wl_listener is is already tagged as freed so unexpected behavior can happens. In any case we were using a memory area already freed.

That is the reason why I moved the callback to an auxiliary structure which is attached to the wl_client_add_destroy_listener and it contains a reference to the ViewBackend. This instance is erased in the destroy notification callback. 

In other case why are going to experiment this crash:


```
#1  0x70683b6c in wl_priv_signal_emit (signal=signal@entry=0x6f3d2650, data=data@entry=0x6f3d2610) at /root/sources/wayland/src/wayland-server.c:2024
#2  0x70683d00 in wl_client_destroy (client=client@entry=0x6f3d2610) at /root/sources/wayland/src/wayland-server.c:866
#3  0x70683e8c in wl_client_connection_data (fd=64, mask=<optimized out>, data=0x6f3d2610) at /root/sources/wayland/src/wayland-server.c:342
#4  0x70685d10 in wl_event_loop_dispatch (loop=0x6f310e38, timeout=timeout@entry=0) at /root/sources/wayland/src/event-loop.c:426
#5  0x6f4c69b8 in WS::ServerSource::<lambda(GSource*, GSourceFunc, gpointer)>::operator() (__closure=0x0, base=0x6f3110f0) at /root/sources/sources/wpebackend-fdo/src/ws.cpp:120
#6  WS::ServerSource::<lambda(GSource*, GSourceFunc, gpointer)>::_FUN(GSource *, GSourceFunc, gpointer) () at /root/sources/sources/wpebackend-fdo/src/ws.cpp:129
#7  0x73f76668 in g_main_context_dispatch () from /root/usr/lib/libglib-2.0.so.0.5800.3
#8  0x73f7688c in ?? () from /root/usr/lib/libglib-2.0.so.0.5800.3
#9  0x73f76c2c in g_main_loop_run () from /root/usr/lib/libglib-2.0.so.0.5800.3
#10 0x004e06f8 in wpe::Launcher::runThread (this=this@entry=0x21b19c0) at /root/sources/sources/wpe-launcher/src/Launcher.cc:895
#11 0x004eb414 in wpe::Launcher::<lambda(void*)>::operator() (__closure=0x0, data=<optimized out>) at /root/sources/sources/wpe-launcher/src/Launcher.cc:499
#12 wpe::Launcher::<lambda(void*)>::_FUN(void *) () at /root/sources/sources/wpe-launcher/src/Launcher.cc:503
#13 0x73fa2bdc in ?? () from /root/usr/lib/libglib-2.0.so.0.5800.3
#14 0x766ebe8c in ?? () from /root/lib/libpthread-2.26.so
```


Where there `wl_priv_signal_emit()` fails for this reason:

```
void
wl_priv_signal_emit(struct wl_priv_signal *signal, void *data)
{
        struct wl_listener *l;
        struct wl_list *pos;

        wl_list_insert_list(&signal->emit_list, &signal->listener_list);
        wl_list_init(&signal->listener_list);

        /* Take every element out of the list and put them in a temporary list.
         * This way, the 'it' func can remove any element it wants from the list
         * without troubles, because we always get the first element, not the
         * one after the current, which may be invalid.
         * wl_list_for_each_safe tries to be safe but it fails: it works fine
         * if the current item is removed, but not if the next one is. */
        while (!wl_list_empty(&signal->emit_list)) {
                pos = signal->emit_list.next;
                l = wl_container_of(pos, l, link);
                fprintf(stderr, "XXX wl_priv_signal_emit: l->notify: %p \n", l->notify);  // <<<<<<< Shows a memory address different to the memory where it should be the struct wl_listener pointer associated to the ViewBackend when this was already deleted

                wl_list_remove(pos);
                wl_list_insert(&signal->listener_list, pos);

                l->notify(l, data);  // <<<<<<< Crash
        }
}
```

Reply:
The destructor for `ViewBackend` needs to do this to unregister the listener for the destruction signal. It is not obvious because “apparently” there is no API for that, but reading the documentation for `wl_signal` one can find this hidden gem (emphasis mine):

> […] Observers are `wl_listener`'s that are added through `wl_signal_add`. Signals are emitted using `wl_signal_emit`, which will invoke all listeners <strong>until that listener is removed by `wl_list_remove()`</strong> (or whenever the signal is destroyed).

So having something like this:

```cpp
class ViewBackend final : // ... 
{
private:
    // ...
    struct wl_listener m_destroyListener { };
};
```

I would actually add the call to remove the listener in `ViewBackend::unregisterSurface()`, which is also called by the `~ViewBackend()` destructor, but can be called before the instance is deleted, in response to an IPC message as well.
 
```cpp
void ViewBackend::unregisterSurface(uint32_t surfaceId)
{
    if (!surfaceId || m_surfaceId != surfaceId)
        return;

    wl_list_remove(&m_destroyListener);  // Added!
    clearFrameCallbacks();

    WS::Instance::singleton().unregisterViewBackend(m_surfaceId);
    m_surfaceId = 0;
}
```

I would initialize `m_destroyListener.notify` inside the class constructor: it can be only once because the function assigned is always the same and it won't change, so there is no need to initialize it every time that `ViewBackend::registerSurface()` is called. The call to register the listener with `wl_client_add_destroy_listener()` would still be done in `ViewBackend::registerSurface()`, of course.

Reply:
`wl_list_remove()`  ... interesting ... actually is what the `wl_priv_signal_emit()` function does into the while. The question  is, that function belongs to the `wayland/src/wayland-private.h` header. I :thinking: if it is good idea to use such type of calls outside the wayland code.

Reply:
Nononono. I am telling you to use `wl_list_remove()`, which is declared is in the **public** `wayland-util.h` header.

Reply:
Ooops. I bad grep-ing before. Yes you are right:


```
../wayland/src/wayland-server-core.h: * If you need to remove a listener from a signal, use wl_list_remove().
../wayland/src/wayland-server-core.h: * wl_list_remove(&your_listener.link);
../wayland/src/wayland-server-core.h: * listener is removed by wl_list_remove() (or whenever the signal is
../wayland/src/wayland-util.h: * safety by invoking wl_list_init() before wl_list_remove().
../wayland/src/wayland-util.h:wl_list_remove(struct wl_list *elm);
```

Looks like a true candidate.